clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -633,77 +633,94 @@ raw2trace_t::do_conversion()
-                                  app_pc block_start, app_pc pc)
+                                  app_pc block_start, int index, app_pc pc)
 {
-    return lookup_instr_summary(tls, modidx, modoffs, block_start, pc) != nullptr;
+    return lookup_instr_summary(tls, modidx, modoffs, block_start, index, pc, nullptr) !=
+        nullptr;
 }
 
-const instr_summary_t *
-raw2trace_t::get_instr_summary(void *tls, uint64 modidx, uint64 modoffs,
-                               app_pc block_start, INOUT app_pc *pc, app_pc orig)
+instr_summary_t *
+raw2trace_t::create_instr_summary(void *tls, uint64 modidx, uint64 modoffs,
+                                  block_summary_t *block, app_pc block_start,
+                                  int instr_count, int index, INOUT app_pc *pc,
+                                  app_pc orig)
 {
     auto tdata = reinterpret_cast<raw2trace_thread_data_t *>(tls);
-    const app_pc decode_pc = *pc;
-    // For rep string loops we expect the same PC many times in a row.
-    if (decode_pc == tdata->last_decode_pc &&
-        block_start == tdata->last_decode_block_start) {
-        *pc = tdata->last_summary->next_pc();
-        return tdata->last_summary;
+    if (block == nullptr) {
+        block = new block_summary_t(block_start, instr_count);
+        DEBUG_ASSERT(index >= 0 && index < static_cast<int>(block->instrs.size()));
+        hashtable_add(&decode_cache[tdata->worker], block_start, block);
+        VPRINT(5, "Created new block summary " PFX " for " PFX "\n", block, block_start);
+        tdata->last_decode_block_start = block_start;
+        tdata->last_block_summary = block;
     }
+    instr_summary_t *desc = &block->instrs[index];
+    if (!instr_summary_t::construct(dcontext, block_start, pc, orig, desc, verbosity)) {
+        WARN("Encountered invalid/undecodable instr @ %s+" PIFX,
+             modvec()[static_cast<size_t>(modidx)].path, IF_NOT_X64((uint)) modoffs);
+        return nullptr;
+    }
+    return desc;
+}
+
+const instr_summary_t *
+raw2trace_t::get_instr_summary(void *tls, uint64 modidx, uint64 modoffs,
+                               app_pc block_start, int instr_count, int index,
+                               INOUT app_pc *pc, app_pc orig)
+{
+    block_summary_t *block;
     const instr_summary_t *ret =
-        lookup_instr_summary(tls, modidx, modoffs, block_start, *pc);
+        lookup_instr_summary(tls, modidx, modoffs, block_start, index, *pc, &block);
     if (ret == nullptr) {
-        instr_summary_t *desc = new instr_summary_t();
-        if (!instr_summary_t::construct(dcontext, block_start, pc, orig, desc,
-                                        verbosity)) {
-            WARN("Encountered invalid/undecodable instr @ %s+" PIFX,
-                 modvec()[static_cast<size_t>(modidx)].path, IF_NOT_X64((uint)) modoffs);
-            return nullptr;
-        }
-        hashtable_add(&decode_cache[tdata->worker], desc, desc);
+        instr_summary_t *desc = create_instr_summary(
```

Comment:
nit: just set 'ret' instead of having a local 'desc' variable?

Reply:
Done.